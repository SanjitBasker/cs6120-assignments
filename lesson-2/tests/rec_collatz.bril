@collatz(n : int) : int {
  print n;

  one : int = const 1;
  two : int = const 2;
  three : int = const 3;

  is_one : bool = eq n one;
  br is_one .is_one .not_one;

  .is_one:
    ret one;
  
  .not_one:
    half : int = div n two;
    round : int= mul half two;
    match : bool = eq round n;
    br match .even .odd;
  
  .even:
    next : int = id half;
    jmp .rec;

  .odd:
    trip_n : int = mul three n;
    next : int = add trip_n one;
  
  .rec:
    res : int = call @collatz next;
    ret res;
}

@main() {
  n : int = const 128;
  r : int = call @collatz n;
}
