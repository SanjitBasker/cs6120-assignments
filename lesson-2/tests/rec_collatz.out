{
  "collatz": [
    {
      "callee_name": "collatz",
      "pos": {
        "col": 5,
        "row": 29
      }
    }
  ],
  "main": [
    {
      "callee_name": "collatz",
      "pos": {
        "col": 3,
        "row": 35
      }
    }
  ]
}
